kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,1338 @@
+* 5. If the `transform` is Index: put the column at the correct location
+* 6. If the `transform` is Cast: cast the column to the specified type, and put it at the correct
+*     location
+* 7. If the `transform` is Missing: put a column of `null` at the correct location
+* 8. If the `transform` is Child([child_order]) and the data is a `StructArray` o, recursively call
+*     `reorder_struct_array` on the column with `child_order` and put the resulting, now correctly
+*     ordered array, at the correct location
+* 9. If the `transform` is Child and the data is a `List<StructArray>`, get the inner struct array
+*     out of the list, reorder it recursively as above, rebuild the list, and the put the column
+*     at the correct location
+*
+* Example:
+* The parquet crate treats columns being actually "flat", so a struct column is purely a schema
+* level thing and doesn't "count" wrt. column indices.
+*
+* So if we have the following file physical schema:
+*
+*  a
+*    d
+*    x
+*  b
+*    y
+*      z
+*    e
+*    f
+*  c
+*
+* and a logical requested schema of:
+*
+*  b
+*    f
+*    e
+*  a
+*    x
+*  c
+*
+* The mask is [1, 3, 4, 5] because a, b, and y don't contribute to the column indices.
+*
+* The reorder tree is:
+* [
+*   // col a is at position 0 in the struct array, and should be moved to position 1
+*   { index: 1, Child([{ index: 0 }]) },
+*   // col b is at position 1 in the struct array, and should be moved to position 0
+*   //   also, the inner struct array needs to be reordered to swap 'f' and 'e'
+*   { index: 0, Child([{ index: 1 }, {index: 0}]) },
+*   // col c is at position 2 in the struct array, and should stay there
+*   { index: 2 }
+* ]
+*/
+
+/// Reordering is specified as a tree. Each level is a vec of `ReorderIndex`s. Each element's index
+/// represents a column that will be in the read parquet data at that level and index. The `index`
+/// of the element is the position that the column should appear in the final output. The `transform`
+/// indicates what, if any, transforms are needed. See the docs for [`ReorderIndexTransform`] for the
+/// meaning.
+#[derive(Debug, PartialEq)]
+pub(crate) struct ReorderIndex {
+    pub(crate) index: usize,
+    transform: ReorderIndexTransform,
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum ReorderIndexTransform {
+    /// For a non-nested type, indicates that we need to cast to the contained type
+    Cast(ArrowDataType),
+    /// Used for struct/list/map. Potentially transform child fields using contained reordering
+    Child(Vec<ReorderIndex>),
+    /// No work needed to transform this data
+    None,
+    /// Data is missing, fill in with a null column
+    Missing(ArrowFieldRef),
+}
+
+impl ReorderIndex {
+    fn new_cast(index: usize, target: ArrowDataType) -> Self {
+        ReorderIndex {
+            index,
+            transform: ReorderIndexTransform::Cast(target),
+        }
```

Comment:
nit: can shave line count a bit by pulling out a value:
```suggestion
        let transform = ReorderIndexTransform::Cast(target);
        ReorderIndex { index, transform }
```
or even add and use a (possibly private) `new` method:
```suggestion
        ReorderIndex::new(index, ReorderIndexTransform::Cast(target))
```

(more below)